File: docs/system-design/01-fundamentals/01-caching/03-application-patterns/01-lru/code/java/Lru.java
import java.util.*;

// LRU cache using LinkedHashMap in access-order.
// Run: javac Lru.java && java Lru
class Lru<K, V> extends LinkedHashMap<K, V> {
    private final int capacity;

    public Lru(int capacity) {
        // true = accessOrder (not insertion order)
        super(Math.max(16, capacity * 2), 0.75f, true);
        if (capacity <= 0) throw new IllegalArgumentException("capacity > 0");
        this.capacity = capacity;
    }

    @Override
    protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
        return size() > capacity;
    }

    public V getValue(K key) {
        return super.getOrDefault(key, null);
    }

    public void putValue(K key, V value) {
        super.put(key, value);
    }

    // Minimal tests
    public static void main(String[] args) {
        Lru<String, Integer> lru = new Lru<>(2);
        lru.putValue("a", 1);
        lru.putValue("b", 2);
        assert Objects.equals(lru.getValue("a"), 1); // a becomes MRU, b is LRU
        lru.putValue("c", 3); // evicts b
        assert lru.getValue("b") == null;
        assert Objects.equals(lru.getValue("a"), 1);
        assert Objects.equals(lru.getValue("c"), 3);
        lru.putValue("d", 4); // evicts a
        assert lru.getValue("a") == null;
        assert Objects.equals(lru.getValue("c"), 3);
        assert Objects.equals(lru.getValue("d"), 4);
        System.out.println("LRU basic tests passed.");
    }
}

File: docs/system-design/01-fundamentals/01-caching/03-application-patterns/01-lru/code/python/lru.py
"""
LRU cache using Python's OrderedDict (O(1) get/put under typical assumptions).

Run: python lru.py
"""
from collections import OrderedDict


class LRU:
    def __init__(self, capacity: int):
        assert capacity > 0
        self.cap = capacity
        self.map: OrderedDict = OrderedDict()

    def get(self, key):
        if key not in self.map:
            return None
        # move to end to mark as recently used
        val = self.map.pop(key)
        self.map[key] = val
        return val

    def put(self, key, value):
        if key in self.map:
            self.map.pop(key)
        elif len(self.map) >= self.cap:
            # popitem(last=False) pops LRU
            self.map.popitem(last=False)
        self.map[key] = value


def _run_basic_tests():
    lru = LRU(2)
    lru.put("a", 1)
    lru.put("b", 2)
    assert lru.get("a") == 1  # a is MRU, b becomes LRU
    lru.put("c", 3)           # evicts b
    assert lru.get("b") is None
    assert lru.get("a") == 1
    assert lru.get("c") == 3
    lru.put("d", 4)           # evicts a
    assert lru.get("a") is None
    assert lru.get("c") == 3
    assert lru.get("d") == 4
    print("LRU basic tests passed.")


if __name__ == "__main__":
    _run_basic_tests()
